watertrading/code/BuildModel.py

```diff
@@ -0,0 +1,1361 @@
+    )
+
+    # set for consumer-node mapping
+    def ConsumerNodeMapINIT(model,c):
+        cies = []
+        for ci in model.s_CI:
+            if Consumer_Wellpad[ci] == c:
+                cies.append(ci)
+        return cies
+    model.ConsumerNodeMap = Set(
+        model.s_CP,
+        dimen = 1,
+        initialize=ConsumerNodeMapINIT,
+        doc="Mapping from consumer node c in s_CP to ci in s_CI (one-to-many)"
+    )
+
+    # set for consumer-node-time mapping
+    def ConsumerNodeTimeMapINIT(model,c,t):
+        ceties = []
+        for ci in model.s_CI:
+            if Consumer_Wellpad[ci] == c and t in model.s_T_ci[ci]:
+                ceties.append(ci)
+        return ceties
+    model.ConsumerNodeTimeMap = Set(
+        model.s_CP,
+        model.s_T,
+        dimen = 1,
+        initialize=ConsumerNodeTimeMapINIT,
+        doc="Mapping from consumer node c in s_CP active at time t in s_T to ci in s_CI (one-to-many)"
+    )
+
+    # set yielding active midstreams by time point
+    def MidstreamTimeMapINIT(model,t):
+        mies = []
+        for mi in model.s_MI:
+            if t in model.s_T_mi[mi]:
+                mies.append(mi)
+        return mies
+    model.MidstreamTimeMap = Set(
+        model.s_T,
+        dimen=1,
+        initialize=MidstreamTimeMapINIT,
+        doc="All midstreams active at time t in s_T (one-to-many)"
+    )
+
+    print("Starting midstream arc set")
+    # set yielding valid time lags for each midstream mi from each start point tp
+    def LLM_INIT(model):
+        LLMs = []
+        for pi in list(model.s_PI.ordered_data()):
+            for ci in list(model.s_CI.ordered_data()):
+                for mi in list(model.s_MI.ordered_data()):
+                    p = model.p_ProducerPad[pi]
+                    c = model.p_ConsumerPad[ci]
+                    if p != c:
+                        for tp in model.s_T_pi[pi]:
+                            for tc in model.s_T_ci[ci]:
+                                if bool(model.MidstreamTimeMap[tp]) and bool(model.ProducerNodeTimeMap[p,tp]) and bool(model.ConsumerNodeTimeMap[c,tc]):
+                                    if model.d_T_ord[tc] - model.d_T_ord[tp] >= 0 and model.d_T_ord[tc] - model.d_T_ord[tp] <= model.p_MidstreamLag[mi].value and model.p_ArcDistance[p,c].value <= model.p_MidstreamMaxdis[mi].value:
+                                        LLMs.append((mi,p,c,tp,tc))
+        """
+        # Works, but slow and inefficient
+        for (mi,p,c,tp,tc) in model.s_MI*model.s_PP*model.s_CP*model.s_T*model.s_T:
+            if bool(model.MidstreamTimeMap[tp]) and bool(model.ProducerNodeTimeMap[p,tp]) and bool(model.ConsumerNodeTimeMap[c,tc]):
+                if model.d_T_ord[tc] - model.d_T_ord[tp] <= model.p_MidstreamLag[mi].value:
+                    LLMs.append((mi,p,c,tp,tc))
+                else: Set.Skip
+            else:
+                Set.Skip
+        """
```

Comment:
I would recommend removing the above commented code if that is not going to be used in future.